File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ArgumentMatcher.java
package io.cucumber.cucumberexpressions;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class ArgumentMatcher {
    static List<Argument> matchArguments(Pattern pattern, String text, List<Transform<?>> transforms) {
        Matcher matcher = pattern.matcher(text);
        if (matcher.lookingAt()) {
            if (matcher.groupCount() != transforms.size()) {
                throw new RuntimeException(String.format("Expression has %s arguments, but there were %s transforms", matcher.groupCount(), transforms.size()));
            }
            List<Argument> arguments = new ArrayList<>(matcher.groupCount());
            for (int i = 0; i < matcher.groupCount(); i++) {
                int startIndex = matcher.start(i + 1);
                String value = matcher.group(i + 1);

                Transform transform = transforms.get(i);
                Object transformedValue = transform.transform(value);

                arguments.add(new Argument(startIndex, transformedValue));
            }
            return arguments;
        } else {
            return null;
        }

    }
}

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/argument_matcher.rb
require 'cucumber/cucumber_expressions/argument'

module Cucumber
  module CucumberExpressions
    class ArgumentMatcher
      def self.match_arguments(regexp, text, transforms)
        m = regexp.match(text)
        return nil if m.nil?
        (1...m.length).map do |index|
          value = m[index]
          transform = transforms[index-1]
          transformed_value = transform.transform(value)
          Argument.new(m.offset(index)[0], value, transformed_value)
        end
      end
    end
  end
end
